src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -95,39 +96,107 @@ public TagHelperDescriptorProvider(IEnumerable<TagHelperDescriptor> descriptors)
                             RequiredAttributeWildcardSuffix,
                             StringComparison.OrdinalIgnoreCase))
                         {
-                            var prefix = requiredAttribute.Substring(0, requiredAttribute.Length - 1);
-
-                            if (!attributeNames.Any(
-                                attributeName =>
-                                    attributeName.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) &&
-                                    !string.Equals(attributeName, prefix, StringComparison.OrdinalIgnoreCase)))
-                            {
-                                return false;
-                            }
+                            var prefix = attributeName.Substring(0, attributeName.Length - RequiredAttributeWildcardSuffix.Length);
+                            matchingAttributes = attributes.Where(
+                                attribute => attribute.Key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) &&
+                                    !string.Equals(attribute.Key, prefix, StringComparison.OrdinalIgnoreCase));
                         }
-                        else if (!attributeNames.Contains(requiredAttribute, StringComparer.OrdinalIgnoreCase))
+                        else
+                        {
+                            matchingAttributes = attributes.Where(
+                                attribute => string.Equals(attribute.Key, attributeName, StringComparison.OrdinalIgnoreCase));
+                        }
+
+                        if (!matchingAttributes.Any())
                         {
                             return false;
                         }
+
+                        // At least one attribute where a name matches, now match the value (if restricted).
+                        if (isSelector)
+                        {
+                            var equalsIndex = requiredAttribute.IndexOf('=');
+                            var afterEquals = requiredAttribute[equalsIndex + 1];
+                            var valueOffset = 0;
+                            if (afterEquals == '\'' || afterEquals == '"')
+                            {
+                                valueOffset = 1;
+                            }
+                            var valueStart = equalsIndex + 1 + valueOffset;
+                            var valueLength = requiredAttribute.Length - equalsIndex - valueOffset * 2 - 2;
+                            var requiredValue = requiredAttribute.Substring(valueStart, valueLength);
+                            var selectorCharacter = requiredAttribute[equalsIndex - 1];
+
+                            switch (selectorCharacter)
+                            {
+                                case '^': // Value starts with
+                                    if (!matchingAttributes.Any(attribute => attribute.Value.StartsWith(requiredValue, StringComparison.Ordinal)))
```

Comment:
Protecting the user from errors should be fine here. Much like how we don't support `*=` `~=` etc. Better to just notify the user they're doing something that doesn't make sense in TagHelper landscape.